### ISSUE TITLE
add methods to extract node information for individuals

### ISSUE THREAD

**Issue by petrelharp** [tskit-dev/tskit#1481]
`time` and `population` is stored by *node* even though it is an individual property. (at least, in all our use cases so far...) So, it's helpful to be able to pull this information out for individuals from their nodes, and it's a bit annoying to do that efficiently. I've got pyslim pre-extracting this information [here](https://github.com/tskit-dev/pyslim/blob/ced273690ca930ef27f6efa205ea20ff419e7361/pyslim/slim_tree_sequence.py#L180) (and, checking that there's no conflicting information). I think this should probably be a tskit method (and, writing it in C would make efficient error checking much easier than what I did there =).

So, I'm proposing
```
class TreeSequence:
   def individual_times(self):
       """
       Returns an array of times of length equal to the number of individuals, so that all nodes of individual ``i``
       have ``time`` attribute equal to the ``i``-the entry of the returned array. Raises an error if the nodes of an individual
       do not all have the same time.
       """

   def individual_populations(self):
```

I suppose I should provide an example showing that doing this in python in the obvious way can take a lot of time, e.g.
```
individual_times = [ts.node(ind.nodes[0]).time if len(ind.nodes) > 0 else np.nan for ind in ts.individuals()]
```
... I swear, in applications it has taken a while.

I guess alternatively we could make these properties of the Individual, so if we wanted a numpy array we'd do
```
individual_times = np.array([ind.time for ind in ts.individuals()])
```

**Comment by jeromekelleher** [comment]
SGTM. I think we should also make ``time`` and ``population`` properties of the Individual class, which will raise an error if the times aren't the same in all nodes. 

I think it makes sense to have them on the TreeSequence rather than the TableCollection.

**Comment by benjeffery** [comment]
Yes, these can be analogous to `site.mutations`, in that they are conveniences in the `TreeSequence` entity classes. Having them in `TableCollection` is bug-prone as they can get out of sync.

They could be lazily created if expensive.

**Comment by molpopgen** [comment]
This is useful.  I've kept the node info as individual metadata in order to do this kinda thing.

**Comment by petrelharp** [comment]
I'm removing `SlimTreeSequence` from pyslim, which was providing this functionality there. I'd like to get this in to tskit so we won't have (much) breakage. I'll make up a PR.

**Comment by petrelharp** [comment]
Proposal: also provide `individual_locations`, which is a numpy array with one row per individual, as long as everyone has the same length of location (and error otherwise).

**Comment by jeromekelleher** [comment]
SGTM.

**Comment by petrelharp** [comment]
> I think we should also make time and population properties of the Individual class, which will raise an error if the times aren't the same in all nodes.

I think so too, but I'm not sure how this would work with `print( )`? Namely, if someone does `print(ts.individual(0))` but individual 0 has mismatching node times, it should not throw an error. We could either
- just never print out times and locations, or
- print out some default value (`None`?) when these produce an error, or
- don't print times and locations only when these produce an error?

I'm happy to do this (esp if we go with option (1) =).

Also, for the record, I don't think we should return a list of times or populations when there's more than one is good, as then we'd want to *always* return lists, mostly of length 1. Folks can get that information from `ind.nodes` anyhow, and in the mismatching case these are not really properties of the individual anyhow.

**Comment by jeromekelleher** [comment]
> I'm happy to do this (esp if we go with option (1) =).

How about we do (1) for now, and make an issue tracking the semantics we might like to have otherwise?

> Also, for the record, I don't think we should return a list of times or populations when there's more than one is good, as then we'd want to always return lists, mostly of length 1. Folks can get that information from ind.nodes anyhow, and in the mismatching case these are not really properties of the individual anyhow.

+1

**Comment by petrelharp** [comment]
Oh, duh - I think (1) makes sense since it should probably just be a `@property`; can do it all in python, even.

**Comment by jeromekelleher** [comment]
Closed in #2305
